### PR TITLE
Issue tests #200: Spectrum tests and associated changes to bin-related methods and docstrings

### PR DESCRIPTION
**Pull request by jvavrek**
This splits up a large block of tests I had written previously into several smaller tests focusing separately on list mode and find index functionalities.

### CODE REVIEW

**Comment by jvavrek** [comment]
TODO:
- [ ] update tests once #203 is merged

File: tests/spectrum_test.py [hunk 1]
```diff
@@ -231,58 +231,115 @@ def test_negative_input(spec_data):
...
-    # additional type checking
+    assert len(spec) == NBINS
+    assert np.isclose(widths[0], BW)
+    assert edges[0] == XMIN
+    assert edges[-1] == XMAX
+    assert len(edges) == NBINS + 1
+    assert spec.has_uniform_bins(use_kev=use_kev)
+
+
+@pytest.mark.parametrize('spec', [
```
Comment (markbandstra) [inline tests/spectrum_test.py]: Test calibrated and uncalibrated spectra with different binnings

File: tests/spectrum_test.py [hunk 2]
```diff
@@ -231,58 +231,115 @@ def test_negative_input(spec_data):
...
+        spec.find_bin_index([XMIN], use_kev=False), np.ndarray)
+
+
+@pytest.mark.parametrize('spec', [
+    make_spec_listmode('uniform'),
+    make_spec_listmode('log'),
+    make_spec('uncal'),
+    make_spec('cal')])
+def test_index_out_of_bounds(spec):
+    '''Raise a SpectrumError when we look for a bin index out of bounds.'''
```
Comment (markbandstra) [inline tests/spectrum_test.py]: PEP 257 :)

File: tests/spectrum_test.py (hunk 1 above)
Comment (jvavrek) [reply tests/spectrum_test.py]: It looks like `has_uniform_bins()` doesn't always work with calibrated spectra—it tests if all the widths are the same to within `EPS = np.finfo(float).eps` (`2.220e-16` on my machine) but the calibration multiplication can introduce slightly higher deviations.

Acceptable to test against, say, `2*EPS` ?

Comment (jvavrek) [reply tests/spectrum_test.py]: Actually, more like `100*EPS` might be required...

**Comment by jvavrek** [comment]
I think this is ready to go

**Comment by jccurtis** [comment]
did you update the tests based on #203 as noted above?

File: becquerel/core/spectrum.py [hunk 3]
```diff
@@ -929,51 +929,62 @@ def has_uniform_bins(self, use_kev=True):
...
-                                'uncalibrated Spectrum.')
+            raise UncalibratedError('Cannot access energy bins with an ' +
+                                    'uncalibrated Spectrum.')
 
         bin_widths = self.bin_widths_kev if use_kev else self.bin_widths_raw
         iterator = iter(bin_widths)
         x0 = next(iterator, None)
         for x in iterator:
-            if abs(x/x0 - 1.0) > EPS:
+            if abs(x/x0 - 1.0) > rtol:
```
Comment (jccurtis) [inline becquerel/core/spectrum.py]: why not `np.allclose(x, x[0], rtol=rtol)`?

**Comment by jccurtis** [comment]
Nit: update title to reflect changes to `Spectrum`

File: becquerel/core/spectrum.py (hunk 3 above)
Comment (jvavrek) [reply becquerel/core/spectrum.py]: I found this nice solution [here](https://github.com/0cjs/py-allthesame#8-writing-the-code-the-function), but if you prefer the `np` solution for readability I can change it. The current version has the advantage of terminating early but I can't imagine that mattering much in an absolute sense.

Comment (jvavrek) [reply becquerel/core/spectrum.py]: Quick tests indicate that the `np` solution is substantially faster for all but the shortest of actually-uniform arrays, starting around 70 elements, while the `iter` solution is substantially faster for most non-uniform arrays such as log bins. Either way, everything is < 100 us.

Comment (jvavrek) [reply becquerel/core/spectrum.py]: added a note in the latest commit

**Pull request merged**